Tests/SwiftFormatRulesTests/NoAssignmentInExpressionsTests.swift

```diff
@@ -0,0 +1,110 @@
+import SwiftFormatRules
+
+final class NoAssignmentInExpressionsTests: LintOrFormatRuleTestCase {
+  func testAssignmentInExpressionContextIsDiagnosed() {
+    XCTAssertFormatting(
+      NoAssignmentInExpressions.self,
+      input: """
+        foo(bar, baz = quux, a + b)
+        """,
+      expected: """
+        foo(bar, baz = quux, a + b)
+        """
+    )
+    XCTAssertDiagnosed(.moveAssignmentToOwnStatement, line: 1, column: 10)
+    // Make sure no other expressions were diagnosed.
+    XCTAssertNotDiagnosed(.moveAssignmentToOwnStatement)
+  }
+
+  func testReturnStatementWithoutExpressionIsUnchanged() {
+    XCTAssertFormatting(
+      NoAssignmentInExpressions.self,
+      input: """
+        func foo() {
+          return
+        }
+        """,
+      expected: """
+        func foo() {
+          return
+        }
+        """
+    )
+    XCTAssertNotDiagnosed(.moveAssignmentToOwnStatement)
+  }
+
+  func testReturnStatementWithNonAssignmentExpressionIsUnchanged() {
+    XCTAssertFormatting(
+      NoAssignmentInExpressions.self,
+      input: """
+        func foo() {
+          return a + b
+        }
+        """,
+      expected: """
+        func foo() {
+          return a + b
+        }
+        """
+    )
+    XCTAssertNotDiagnosed(.moveAssignmentToOwnStatement)
+  }
+
+  func testReturnStatementWithSimpleAssignmentExpressionIsExpanded() {
+    // For this and similar tests below, we don't try to match the leading indentation in the new
+    // `return` statement; the pretty-printer will fix it up.
+    XCTAssertFormatting(
+      NoAssignmentInExpressions.self,
+      input: """
+        func foo() {
+          return a = b
+        }
+        """,
+      expected: """
+        func foo() {
+          a = b
+        return
+        }
+        """
+    )
+    XCTAssertDiagnosed(.moveAssignmentToOwnStatement, line: 2, column: 10)
+  }
+
+  func testReturnStatementWithCompoundAssignmentExpressionIsExpanded() {
+    XCTAssertFormatting(
+      NoAssignmentInExpressions.self,
+      input: """
+        func foo() {
+          return a += b
+        }
+        """,
+      expected: """
+        func foo() {
+          a += b
+        return
+        }
+        """
+    )
+    XCTAssertDiagnosed(.moveAssignmentToOwnStatement, line: 2, column: 10)
+  }
+
+  func testReturnStatementWithAssignmentHasCommentMoved() {
```

Comment:
What about if they're doing something _silly_ like opening a multiline comment on the line with the return? I think the trailing trivia will get moved so that the return is sitting inside of the comment?

```
func foo() {
  return a = b /* blargh
    continues here */
}
```

Reply:
The trivia will never break within a block comment, so the entire comment (including its line break) will be a single trivia piece. We're also still using the ~~hack~~ workaround I wrote to preserve the legacy trivia behavior, so that comment would be in the leading trivia of the `}` token.

I did tweak this logic though to make sure it handles more cases around weird comments and added tests for those, PTAL.